# Rare variant analysis for pangenome-aware DeepVariant (population allele frequency < 0.01)

Some reviewers criticized that pangenome-aware DV is biased toward the pangenome that it takes as input, and that the model is skeptical of true rare variants absent from the pangenome. 
Here, we want to show that pangenome-aware DV and linear-reference-based DV are comparable in terms of calling rare variants.
Here are the data we need for this analysis:
- The variant call sets generated by running linear-reference-based DV on 1KG samples (no pangenome is used)
- The variant call sets generated by running pangenome-aware DV on 1KG samples (HPRC graph-v1.1 with all 88 haplotypes is used)
- Five samples were selected from the HPRC Release 2 that were absent from HPRC Release 1 (or absent from graph-v1.1)
- The hifiasm haplotype-resolved assemblies were mapped to GRCh38, and then Dipcall was used to create truth sets for those five HPRC Release 2 samples

Using these data, we will perform the following steps:

- Take all the 1KG linear-reference-based DV VCF files
- Compute the frequency of all variants in the set and select the rare variants; essentially, we use linear-reference-based DV results as a proxy to measure variant rarity
- Filter out rare variants not supported by assembly-based Dipcall call sets. This filtering may not be easy if we include indels and complex regions, but we can use hap.py for this comparison.
  For example, we can pass DV calls as the query and the Dipcall set as the truth and extract only TP calls.
- After obtaining those confident rare variants for the five samples, compare pangenome-aware DV calls against them and report the recall rate.


## Downloading and extracting records for 5 HPRC-R2 samples from 1KG linear-ref-based DV VCF 

We ran linear-ref-based DV on 1KG samples and they were merged into a single VCF file per chromosome. They are available here
```
gsutil ls gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/

gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr1.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr10.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr11.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr12.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr13.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr14.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr15.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr16.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr17.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr18.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr19.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr2.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr20.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr21.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr22.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr3.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr4.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr5.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr6.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr7.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr8.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr9.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chrX.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chrY.vcf.gz
```

Since the size of each chromosome-specifc VCF file was huge we process 4 files at a time with the script `run_filter_hom_ref.bash`. It downloads 4 VCF files at a time (by using `xargs -P 4`) and then pass each downloaded VCF file to the script `filter_hom_ref.py`. `filter_hom_ref.py` will parse the input VCF file and extract the records with at least one non-ref allele among 5 HPRC-R2 samples (`HG01255`,`HG02280`,`HG02984`,`HG03831`,`HG04184`). The list of the samples can be passed to `filter_hom_ref.py` using the `--samples` parameter. The output VCF file which is much smaller than the input file will be saved into the path given to `--output`. Both `filter_hom_ref.py` and its bash wrapper script `run_filter_hom_ref.bash` are available under the `scripts/` folder. 

The output vcf files containing non-ref records for the 5 HPRC-R2 samples were then merged into a single vcf file which is available via these s3 links:
```
https://s3-us-west-2.amazonaws.com/human-pangenomics/submissions/1d4e5127-0bd2-4858-baaf-514c724a925a--PANGENOME_AWARE_DEEPVARIANT/1kg_rare_variant_analysis/all_chroms.1kg_cohort.vg.linear_ref_based_dv.grch38.5_HPRC_samples.with_non_ref_allele.sorted.vcf.gz
https://s3-us-west-2.amazonaws.com/human-pangenomics/submissions/1d4e5127-0bd2-4858-baaf-514c724a925a--PANGENOME_AWARE_DEEPVARIANT/1kg_rare_variant_analysis/all_chroms.1kg_cohort.vg.linear_ref_based_dv.grch38.5_HPRC_samples.with_non_ref_allele.sorted.vcf.gz.tbi
```

## Finding rare variants and making sample-specific vcf files

In the vcf file created in the previous step `all_chroms.1kg_cohort.vg.linear_ref_based_dv.grch38.5_HPRC_samples.with_non_ref_allele.sorted.vcf.gz` there is a column named `AF` which contains the population allele frequency based on the original linear-ref-based DV calls for 1KG samples. We wrote a script named `extract_rare_variants_per_sample.py` that uses this column to filter records with at least one allele whose AF is lower than 0.01 (Note that there might be multiple alternative alleles )

 
