# Rare variant analysis for pangenome-aware DeepVariant (population allele frequency < 0.01)

Some reviewers criticized that pangenome-aware DV is biased toward the pangenome that it takes as input, and that the model is skeptical of true rare variants absent from the pangenome. 
Here, we want to show that pangenome-aware DV and linear-reference-based DV are comparable in terms of calling rare variants.
Here are the data we need for this analysis:
- The variant call sets generated by running linear-reference-based DV on 1KG samples (no pangenome is used)
- The variant call sets generated by running pangenome-aware DV on 1KG samples (HPRC graph-v1.1 with all 88 haplotypes is used)
- Five samples were selected from the HPRC Release 2 that were absent from HPRC Release 1 (or absent from graph-v1.1)
- The hifiasm haplotype-resolved assemblies were mapped to GRCh38, and then Dipcall was used to create truth sets for those five HPRC Release 2 samples

Using these data, we will perform the following steps:

- Take all the 1KG linear-reference-based DV VCF files
- Compute the frequency of all variants in the set and select the rare variants; essentially, we use linear-reference-based DV results as a proxy to measure variant rarity
- Filter out rare variants not supported by assembly-based Dipcall call sets. This filtering may not be easy if we include indels and complex regions, but we can use hap.py for this comparison.
  For example, we can pass DV calls as the query and the Dipcall set as the truth and extract only TP calls.
- After obtaining those confident rare variants for the five samples, compare pangenome-aware DV calls against them and report the recall rate.


## Downloading and extracting records for 5 HPRC-R2 samples from 1KG linear-ref-based DV VCF 

We ran linear-ref-based DV on 1KG samples and they were merged into a single VCF file per chromosome. They are available here
```
gsutil ls gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/

gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr1.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr10.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr11.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr12.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr13.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr14.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr15.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr16.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr17.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr18.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr19.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr2.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr20.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr21.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr22.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr3.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr4.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr5.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr6.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr7.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr8.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chr9.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chrX.vcf.gz
gs://brain-genomics-public/research/cohort/1KGP/vg/dv_grch38/chrY.vcf.gz
```

Since the size of each chromosome-specifc VCF file was huge we process 4 files at a time with the script `run_filter_hom_ref.bash`. It downloads 4 VCF files at a time (by using `xargs -P 4`) and then pass each downloaded VCF file to the script `filter_hom_ref.py`. `filter_hom_ref.py` will parse the input VCF file and extract the records with at least one non-ref allele among 5 HPRC-R2 samples (`HG01255`,`HG02280`,`HG02984`,`HG03831`,`HG04184`). The list of the samples can be passed to `filter_hom_ref.py` using the `--samples` parameter. The output VCF file which is much smaller than the input file will be saved into the path given to `--output`. Both `filter_hom_ref.py` and its bash wrapper script `run_filter_hom_ref.bash` are available under the `scripts/` folder. 

The output vcf files containing non-ref records for the 5 HPRC-R2 samples were then merged into a single vcf file which is available via these s3 links:
```
https://s3-us-west-2.amazonaws.com/human-pangenomics/submissions/1d4e5127-0bd2-4858-baaf-514c724a925a--PANGENOME_AWARE_DEEPVARIANT/1kg_rare_variant_analysis/all_chroms.1kg_cohort.vg.linear_ref_based_dv.grch38.5_HPRC_samples.with_non_ref_allele.sorted.vcf.gz
https://s3-us-west-2.amazonaws.com/human-pangenomics/submissions/1d4e5127-0bd2-4858-baaf-514c724a925a--PANGENOME_AWARE_DEEPVARIANT/1kg_rare_variant_analysis/all_chroms.1kg_cohort.vg.linear_ref_based_dv.grch38.5_HPRC_samples.with_non_ref_allele.sorted.vcf.gz.tbi
```

## Finding rare variants and making sample-specific vcf files

In the vcf file created in the previous step `all_chroms.1kg_cohort.vg.linear_ref_based_dv.grch38.5_HPRC_samples.with_non_ref_allele.sorted.vcf.gz` there is a column named `AF` which contains the population allele frequency based on the original linear-ref-based DV calls for 1KG samples. We wrote a script named `extract_rare_variants_per_sample.py` that uses this column to keep the records with at least one allele whose AF is lower than 0.01 which was given to the `--af-threshold` parameter (Note that there might be multiple alternative alleles for each record and not all of them are neccessarily related to each sample). We ran the following command for rare variant extraction:

```
mkdir rare_variants
python3 extract_rare_variants_per_sample.py \
    --input all_chroms.1kg_cohort.vg.linear_ref_based_dv.grch38.5_HPRC_samples.with_non_ref_allele.sorted.vcf.gz \
    --output-dir rare_variants \
    --samples HG01255,HG02280,HG02984,HG03831,HG04184 \
    --af-threshold 0.01
```

The output sample-specific vcf files are available in this s3 folder:
```
https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=submissions/1d4e5127-0bd2-4858-baaf-514c724a925a--PANGENOME_AWARE_DEEPVARIANT/1kg_rare_variant_analysis/rare_variants/
```

## Benchmarking rare variants against the assembly-based call sets (created with Dipcall)

In order to make sure that the rare variants detected in the previous step are reliable we benchmark them against the assembly-based call sets that were created by mapping HPRC-R2 hifiasm assemblies to GRCh38 and running Dipcall. We then extracted only the true positives to be used as the reliable calls. The document that describes the generation of assembly-based call sets is available here: https://github.com/mobinasri/research_notes/tree/main/DeepVariant_Pangenome/Assembly_Based_Truth_Set_HPRC_R2 

For benchmarking the rare variants against the assembly-based call sets we used [happy](https://github.com/Illumina/hap.py). The output happy vcf file is then filtered to contain only the TP calls and saved into files with `.TP.vcf.gz` suffix. To be able to use these files later as the truth set for hap.py we removed the `QUERY` column from them and saved into files with files with `TP.TRUTH_only.vcf.gz` suffix. 
```
cd /private/groups/patenlab/masri/haplotype_sampling/pangenome_aware_dv_paper/1KG_analysis/linear_ref_based_dv/5_HPRC_samples/compare_dipcall_truth

for i in HG01255 HG02280 HG02984 HG03831 HG04184;do
    bcftools view \
        -s TRUTH ${i}/${i}_hprc_v2.rare_af_lt_0.01.against_dipcall_truth_set.TP.vcf.gz \
        -Oz \
        -o ${i}/${i}_hprc_v2.rare_af_lt_0.01.against_dipcall_truth_set.TP.TRUTH_only.vcf.gz
    tabix ${i}/${i}_hprc_v2.rare_af_lt_0.01.against_dipcall_truth_set.TP.TRUTH_only.vcf.gz
done
```

The happy files are available in this s3 folder:
```
https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=submissions/1d4e5127-0bd2-4858-baaf-514c724a925a--PANGENOME_AWARE_DEEPVARIANT/1kg_rare_variant_analysis/compare_dipcall_truth/
```

## Benchmarking pangenome-aware DV calls against confident rare variants for computing recall rates

The pangenome-aware DV calls are downloaded via these gs links:
```
cd /private/groups/patenlab/masri/haplotype_sampling/pangenome_aware_dv_paper/1KG_analysis/linear_ref_based_dv/5_HPRC_samples/benchmark_pangenome_aware_dv/vcf_files

for i in HG01255 HG02280 HG02984 HG03831 HG04184;do
gsutil cp gs://brain-genomics-public/research/cohort/1KGP/vg/pangenome_aware_to_grch38_vcf/pangenome_aware_dv_output/$i.vcf.gz .
gsutil cp gs://brain-genomics-public/research/cohort/1KGP/vg/pangenome_aware_to_grch38_vcf/pangenome_aware_dv_output/$i.vcf.gz.tbi .
done
```

Here we make a template input json file with two placeholders `SAMPLE_PLACEHOLDER` and `GENDER_PLACEHOLDER`
```
{
	"truth_vcf" : "/private/groups/patenlab/masri/haplotype_sampling/pangenome_aware_dv_paper/1KG_analysis/linear_ref_based_dv/5_HPRC_samples/compare_dipcall_truth/SAMPLE_PLACEHOLDER/SAMPLE_PLACEHOLDER_hprc_v2.rare_af_lt_0.01.against_dipcall_truth_set.TP.TRUTH_only.vcf.gz", 
	"truth_conf_bed" : "/private/groups/patenlab/masri/internship/assembly_truth_sets/dipcall_results_GRCh38/SAMPLE_PLACEHOLDER/SAMPLE_PLACEHOLDER_hprc_r2_v1.0.1_dipcall_GRCh38.dip.bed",
	"query_vcf" : "/private/groups/patenlab/masri/haplotype_sampling/pangenome_aware_dv_paper/1KG_analysis/linear_ref_based_dv/5_HPRC_samples/benchmark_pangenome_aware_dv/vcf_files/SAMPLE_PLACEHOLDER.vcf.gz",
	"reference_fasta" : "/private/groups/patenlab/masri/internship/assembly_truth_sets/GRCh38/hs38.fa",
	"output_prefix" : "SAMPLE_PLACEHOLDER.pangenome_aware_dv_vs_conf_rare",
	"output_dir" : "/private/groups/patenlab/masri/haplotype_sampling/pangenome_aware_dv_paper/1KG_analysis/linear_ref_based_dv/5_HPRC_samples/",
	"stratification_tsv" : "",
	"mount_dir" : "/private/groups/patenlab/masri",
	"threads" : "16",
	"gender" : "GENDER_PLACEHOLDER"
}
```


Make input json files:
```
cd /private/groups/patenlab/masri/haplotype_sampling/pangenome_aware_dv_paper/1KG_analysis/linear_ref_based_dv/5_HPRC_samples/benchmark_pangenome_aware_dv

mkdir -p input_jsons
declare -A sampletosex=( ["HG01255"]="male" ["HG02280"]="female"  ["HG02984"]="male"  ["HG03831"]="female"  ["HG04184"]="female" )

for i in HG01255 HG02280 HG02984 HG03831 HG04184;do
  sed -e s@SAMPLE_PLACEHOLDER@$i@g  -e s@GENDER_PLACEHOLDER@${sampletosex[$i]}@g happy.inputs.json > input_jsons/happy.inputs.$i.json
  mkdir -p $i
done
```

Run hap.py
```
for i in HG01255 HG02280 HG02984 HG03831 HG04184;do
  sbatch /private/groups/patenlab/masri/apps/bash_scripts/happy.bash input_jsons/happy.inputs.$i.json
done

```

## Results

I just report the part related to recall rates since precision and f1-scores does not mean anything in this analysis. The recall rate is between 95% to 99% which means that the majority of 
the confident rare variants detected by linear-ref-based DV are being called by pangenome-aware DV so using pangenome is not affecting its performance on rare variants.


### HG01255
| Type  | Filter | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ----- | ------ | ----------- | -------- | -------- | ------------- |
| INDEL | ALL    | 19249       | 18424    | 825      | 0.957141      |
| SNP   | ALL    | 103628      | 102199   | 1429     | 0.98621       |



### HG02280
| Type  | Filter | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ----- | ------ | ----------- | -------- | -------- | ------------- |
| INDEL | ALL    | 54276       | 52654    | 1622     | 0.970116      |
| SNP   | ALL    | 272294      | 269004   | 3290     | 0.987917      |



### HG02984
| Type  | Filter | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ----- | ------ | ----------- | -------- | -------- | ------------- |
| INDEL | ALL    | 54979       | 53139    | 1840     | 0.966533      |
| SNP   | ALL    | 264986      | 261961   | 3025     | 0.988584      |


### HG03831
| Type  | Filter | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ----- | ------ | ----------- | -------- | -------- | ------------- |
| INDEL | ALL    | 21184       | 20349    | 835      | 0.960583      |
| SNP   | ALL    | 118226      | 116193   | 2033     | 0.982804      |


### HG04184
| Type  | Filter | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ----- | ------ | ----------- | -------- | -------- | ------------- |
| INDEL | ALL    | 21358       | 20401    | 957      | 0.955192      |
| SNP   | ALL    | 118963      | 116447   | 2516     | 0.978851      |


## An alternative approach (Compare everything to linear-ref-based DV)

Here, we use the linear-reference-based DV call set as the truth and compare the pangenome-aware DV and Dipcall-based call sets against this truth set. We assume that the Dipcall-based call set is more reliable since it is derived from hifiasm assemblies. Our initial goal was to assess whether the recall rates for pangenome-aware DV are comparable to those of the Dipcall-based call set. The results show that the recall rate for InDels are higher for pangenome-aware DV than for Dipcall. For SNPs they are more comparable. This indicates that we are not losing sensitivity in calling rare variants in DeepVariant by using a pangenome. It also suggests that some InDel errors are shared between the two DV call sets regardless of the use of a pangenome, which is expected since both rely on short-read data and are inherently less accurate than assembly-based approaches.

### HG01255

| Method             | Type  | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ------------------ | ----- | ----------- | -------- | -------- | ------------- |
| pangenome-aware DV | INDEL | 24847       | 22179    | 2668     | 0.892623      |
| pangenome-aware DV | SNP   | 109347      | 105581   | 3766     | 0.965559      |
| Dipcall            | INDEL | 24841       | 19027    | 5814     | 0.765951      |
| Dipcall            | SNP   | 109331      | 102974   | 6357     | 0.941855      |


### HG02280

| Method             | Type  | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ------------------ | ----- | ----------- | -------- | -------- | ------------- |
| pangenome-aware DV | INDEL | 63309       | 58865    | 4444     | 0.929805      |
| pangenome-aware DV | SNP   | 275873      | 269622   | 6251     | 0.977341      |
| Dipcall            | INDEL | 63301       | 54848    | 8453     | 0.866463      |
| Dipcall            | SNP   | 275828      | 270891   | 4937     | 0.982101      |

### HG02984

| Method             | Type  | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ------------------ | ----- | ----------- | -------- | -------- | ------------- |
| pangenome-aware DV | INDEL | 66423       | 61495    | 4928     | 0.925809      |
| pangenome-aware DV | SNP   | 278522      | 271615   | 6907     | 0.975201      |
| Dipcall            | INDEL | 66416       | 55349    | 11067    | 0.833368      |
| Dipcall            | SNP   | 278466      | 263658   | 14808    | 0.946823      |

### HG03831

| Method             | Type  | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ------------------ | ----- | ----------- | -------- | -------- | ------------- |
| pangenome-aware DV | INDEL | 26851       | 24387    | 2464     | 0.908234      |
| pangenome-aware DV | SNP   | 120152      | 116704   | 3448     | 0.971303      |
| Dipcall            | INDEL | 26846       | 21143    | 5703     | 0.787566      |
| Dipcall            | SNP   | 120142      | 117540   | 2602     | 0.978342      |


### HG04184

| Method             | Type  | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | METRIC.Recall |
| ------------------ | ----- | ----------- | -------- | -------- | ------------- |
| pangenome-aware DV | INDEL | 27535       | 24752    | 2783     | 0.898929      |
| pangenome-aware DV | SNP   | 121583      | 117147   | 4436     | 0.963515      |
| Dipcall            | INDEL | 27533       | 21361    | 6172     | 0.775833      |
| Dipcall            | SNP   | 121576      | 118317   | 3259     | 0.973194      |


